transform_aes.c

```diff
@@ -325,36 +338,34 @@ static int transop_decode_aes( n2n_trans_op_t * arg,
 
                 traceEvent( TRACE_DEBUG, "decode_aes %lu with SA %lu.", in_len, sa_rx, sa->sa_id );
 
-                len = (in_len - (TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE));
+                len = (in_len - (TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE + N2N_AES_IVEC_SIZE));
                 
                 if ( 0 == (len % AES_BLOCK_SIZE ) )
                 {
-                    uint8_t padding;
-
+                    uint16_t data_len;
                     memset( &(sa->dec_ivec), 0, sizeof(N2N_AES_IVEC_SIZE) );
-                    AES_cbc_encrypt( (inbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE),
+                   AES_decrypt (inbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE,
+                                 &(sa->dec_ivec),
+                                 &(sa->dec_key_2) );
+
+                    AES_cbc_encrypt( (inbuf + TRANSOP_AES_VER_SIZE + TRANSOP_AES_SA_SIZE + N2N_AES_MSGAUTH_SIZE + N2N_AES_IVEC_SIZE),
                                      assembly, /* destination */
                                      len, 
                                      &(sa->dec_key),
-                                     sa->dec_ivec, 0 /* decrypt */ );
-
-                    /* last byte is how much was padding: max value should be
-                     * AES_BLOCKSIZE-1 */
-                    padding = assembly[ len-1 ] & 0xff; 
+                                     &(sa->dec_ivec), 0 /* decrypt */ );
+                    data_len = *(uint16_t*)assembly;
```

Comment:
For the endianess, replace with: `data_len = ntohs(*(uint16_t*)assembly;)`

Reply:
Thank you!